File: backend/src/main/java/com/pikapikamatch/service/StatsService.java
package com.pikapikamatch.service;

import com.pikapikamatch.exception.ExternalApiException;
import com.pikapikamatch.model.dto.CharacterStatsDTO;
import com.pikapikamatch.model.entity.Character;
import com.pikapikamatch.repository.CharacterRepository;
import com.pikapikamatch.util.CharacterMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Service for managing character statistics operations.
 * Provides methods to query rankings and statistics of characters
 * based on likes and dislikes.
 */
@Service
@Slf4j
@RequiredArgsConstructor
public class StatsService {

    private final CharacterRepository characterRepository;

    /**
     * Gets the character with the most likes.
     *
     * @return CharacterStatsDTO with the most liked character
     * @throws ExternalApiException if no characters exist in the database
     */
    public CharacterStatsDTO getMostLiked() {
        log.debug("Fetching most liked character");
        
        Character mostLiked = characterRepository.findTopByOrderByTotalLikesDesc();
        
        if (mostLiked == null) {
            log.warn("No characters found in database");
            throw new ExternalApiException("No characters available");
        }
        
        log.info("Most liked character: {} with {} likes", 
            mostLiked.getName(), mostLiked.getTotalLikes());
        
        return CharacterMapper.toStatsDTO(mostLiked);
    }

    /**
     * Gets the character with the most dislikes.
     *
     * @return CharacterStatsDTO with the most disliked character
     * @throws ExternalApiException if no characters exist in the database
     */
    public CharacterStatsDTO getMostDisliked() {
        log.debug("Fetching most disliked character");
        
        Character mostDisliked = characterRepository.findTopByOrderByTotalDislikesDesc();
        
        if (mostDisliked == null) {
            log.warn("No characters found in database");
            throw new ExternalApiException("No characters available");
        }
        
        log.info("Most disliked character: {} with {} dislikes", 
            mostDisliked.getName(), mostDisliked.getTotalDislikes());
        
        return CharacterMapper.toStatsDTO(mostDisliked);
    }

    /**
     * Gets the top N characters with the most likes.
     *
     * @param limit The maximum number of characters to return (default: 5)
     * @return List of CharacterStatsDTO sorted by likes in descending order
     */
    public List<CharacterStatsDTO> getTopLiked(Integer limit) {
        log.debug("Fetching top {} liked characters", limit);
        
        List<Character> topLiked = characterRepository.findTopByOrderByTotalLikesDesc(limit);
        
        log.info("Found {} top liked characters", topLiked.size());
        
        return topLiked.stream()
            .map(CharacterMapper::toStatsDTO)
            .collect(Collectors.toList());
    }

    /**
     * Gets the top N characters with the most dislikes.
     *
     * @param limit The maximum number of characters to return (default: 5)
     * @return List of CharacterStatsDTO sorted by dislikes in descending order
     */
    public List<CharacterStatsDTO> getTopDisliked(Integer limit) {
        log.debug("Fetching top {} disliked characters", limit);
        
        List<Character> topDisliked = characterRepository.findTopByOrderByTotalDislikesDesc(limit);
        
        log.info("Found {} top disliked characters", topDisliked.size());
        
        return topDisliked.stream()
            .map(CharacterMapper::toStatsDTO)
            .collect(Collectors.toList());
    }
}

File: front/src/services/statsService.ts
import { apiClient } from './api/apiClient';
import { API_ENDPOINTS } from './api/config';
import type { CharacterStatsDTO } from './api/types';
import type { Character, CharacterStats } from '../types';
import { toCharacterWithStats } from '../utils/mappers';

/**
 * Stats Service
 * Handles all statistics-related API operations
 */
export const statsService = {
  /**
   * Get top liked characters
   * Always fetches fresh data from backend
   * @param limit - Maximum number of characters to retrieve (default: 5)
   * @returns Promise<Array<{ character: Character; stats: CharacterStats }>> - Top liked characters with stats
   */
  async getTopLiked(
    limit: number = 5
  ): Promise<Array<{ character: Character; stats: CharacterStats }>> {
    console.log('[statsService] Fetching top liked from backend');
    try {
      const dtos = await apiClient.get<CharacterStatsDTO[]>(
        API_ENDPOINTS.stats.topLiked,
        { limit }
      );

      return dtos.map(toCharacterWithStats);
    } catch (error: any) {
      // Handle 404 gracefully - no data available yet
      if (error.statusCode === 404) {
        console.log('[statsService] No top liked data available (404)');
        return [];
      }
      throw error;
    }
  },

  /**
   * Get top disliked characters
   * Always fetches fresh data from backend
   * @param limit - Maximum number of characters to retrieve (default: 5)
   * @returns Promise<Array<{ character: Character; stats: CharacterStats }>> - Top disliked characters with stats
   */
  async getTopDisliked(
    limit: number = 5
  ): Promise<Array<{ character: Character; stats: CharacterStats }>> {
    console.log('[statsService] Fetching top disliked from backend');
    try {
      const dtos = await apiClient.get<CharacterStatsDTO[]>(
        API_ENDPOINTS.stats.topDisliked,
        { limit }
      );

      return dtos.map(toCharacterWithStats);
    } catch (error: any) {
      // Handle 404 gracefully - no data available yet
      if (error.statusCode === 404) {
        console.log('[statsService] No top disliked data available (404)');
        return [];
      }
      throw error;
    }
  },

  /**
   * Get most liked character
   * @returns Promise<{ character: Character; stats: CharacterStats } | null> - Most liked character or null if none exists
   */
  async getMostLiked(): Promise<{ character: Character; stats: CharacterStats } | null> {
    console.log('[statsService] Fetching most liked from backend');
    try {
      const dto = await apiClient.get<CharacterStatsDTO>(
        API_ENDPOINTS.stats.mostLiked
      );
      return toCharacterWithStats(dto);
    } catch (error: any) {
      // Handle 404 gracefully - no data available yet
      if (error.statusCode === 404) {
        console.log('[statsService] No most liked data available (404)');
        return null;
      }
      throw error;
    }
  },

  /**
   * Get most disliked character
   * @returns Promise<{ character: Character; stats: CharacterStats } | null> - Most disliked character or null if none exists
   */
  async getMostDisliked(): Promise<{ character: Character; stats: CharacterStats } | null> {
    console.log('[statsService] Fetching most disliked from backend');
    try {
      const dto = await apiClient.get<CharacterStatsDTO>(
        API_ENDPOINTS.stats.mostDisliked
      );
      return toCharacterWithStats(dto);
    } catch (error: any) {
      // Handle 404 gracefully - no data available yet
      if (error.statusCode === 404) {
        console.log('[statsService] No most disliked data available (404)');
        return null;
      }
      throw error;
    }
  },
};
